include/sys/trace_acl.h

```diff
@@ -153,4 +154,10 @@ DEFINE_ACE_EVENT(zfs_zfs__ace__allows);
 #define	TRACE_INCLUDE_FILE trace_acl
 #include <trace/define_trace.h>
 
-#endif /* _KERNEL && HAVE_DECLARE_EVENT_CLASS */
+#else
+
+DEFINE_DTRACE_PROBE3(zfs__ace__denies);
+DEFINE_DTRACE_PROBE3(zfs__ace__allows);
```

Comment:
It looks like every tracepoint is defined both in a HAVE_DECLARE_EVENT_CLASS way, and without that.  Is the rule for adding tracepoints that you need to add a fancy (EVENT_CLASS) declaration, with the DEFINE_DTRACE_PROBE* as a backup?  If so, why does DEFINE_DTRACE_PROBE* handle the case where we HAVE_DECLARE_EVENT_CLASS?

Reply:
There are really three things that need to be specified by the macros:  
1) Calls to the probe functions in the zfs code. 
2) Prototypes for the probe functions. 
3) Probe functions. 

As I understand it, with tracepoints the EVENT_CLASS declarations handle 2 and 3 and the  DEFINE_DTRACE_PROBE* are used for 1.

Reply:
@ahrens, as you mentioned privately  DEFINE_DTRACE_PROBE is not defined for the HAVE_DECLARE_EVENT_CLASS, only when tracepoints are not used.   However; this exchange shows that the code is confusing.  I updated the comments to try and provide some help.